common/src/main/java/com/wynntils/features/user/ChatMentionFeature.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.user;
+
+import com.wynntils.core.config.Config;
+import com.wynntils.core.config.ConfigHolder;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.handlers.chat.event.ChatMessageReceivedEvent;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.mc.McUtils;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class ChatMentionFeature extends UserFeature {
+    @Config
+    public boolean markMention = true;
+
+    @Config
+    public boolean dingMention = true;
+
+    @Config
+    ChatFormatting rewriteColorCode = ChatFormatting.YELLOW;
+
+    @Config
+    public String aliases = "";
+
+    private Pattern pattern;
+
+    @Override
+    protected void onConfigUpdate(ConfigHolder configHolder) {
+        // rebuild pattern in case it has changed
+        buildPattern();
+    }
+
+    private void buildPattern() {
+        pattern = Pattern.compile(
+                "(?<!\\[)\\b(" + McUtils.mc().getUser().getName()
+                        + (aliases.length() > 0 ? "|" + aliases.replace(",", "|") : "") + ")\\b(?!:|])",
+                Pattern.CASE_INSENSITIVE);
+    }
+
+    @SubscribeEvent()
+    public void onChat(ChatMessageReceivedEvent e) {
+        Component message = e.getMessage();
+
+        Matcher looseMatcher = pattern.matcher(ComponentUtils.getUnformatted(message));
+
+        if (looseMatcher.find()) {
+            if (markMention) {
+                e.setMessage(rewriteComponentTree(message));
+            }
+            if (dingMention) {
+                McUtils.playSound(SoundEvents.NOTE_BLOCK_PLING.value());
+            }
+        }
+    }
+
+    private Component rewriteComponentTree(Component comp) {
+        // Make a copy of the component without the siblings
+        MutableComponent curr = MutableComponent.create(comp.getContents()).withStyle(comp.getStyle());
+        // .getString() is used here as it gives formattingchars when those exist. It is needed for guild messages
+        // because wynn still uses legacy coloring for it.
+        String text = curr.getString();
+
+        // if current component has text check it for mentions
+        if (text != "") {
+            Matcher match = pattern.matcher(text);
+
+            // if we match then rewrite the component
+            if (match.find()) {
+                // get the start of the string before any mention and set curr to be it
```

Comment:
Ok, the code is starting to look comprehensive now. The `rewriteComponentTree` method still needs some work though. It is a complex function; that is okay, it is doing a complex thing, but it is also not designed to handle that complexity. 

More to the point: This part, when you have found a match and processes the component, can -- and should! -- be extracted into a separate method.

A good rule of thumb is a method should not be longer than it fits on a single screen. It is not always possible to follow, but here it definitely is. If you extract the "replace text in this component" into a function, you get one screen with the overview of rewriteComponentTree, on a high level, traversing the Component tree and replacing the text when a match is found, and then you can have a `rewriteComponentNode` (or whatever) that focuses on the one single objective of replacing the text in that particular node.

Reply:
My patience is starting to run dry trying to simplify this function more and more. In my opinion its already clear enough what it does. It has a decent name, there are comments explaining what it does, it only has a few major code paths that can occur.

I'll see if I have the patience to try clean it up further tomorrow but at this point I'm heavily considering just abandoning this in hopes that someone else might come along and finish the job.

Reply:
I am sorry you feel that way. Maybe we should have communicated better at the start that the PR needed some heavy rework to reach the quality levels needed for integration.